Add missing `serial` in `NATIVE_DATABASE_TYPES`

kamipo opened the pull request:
- Add missing `serial` type in `NATIVE_DATABASE_TYPES` for PG adapter
  - The short-hand methods was added by #19030.
  - Schema dumping support was added by #19185.
- Add missing geometric types in `NATIVE_DATABASE_TYPES` for PG adapter
  - The short-hand methods was added by #20459.
- Add missing `bigint` type in `NATIVE_DATABASE_TYPES` for SQLite3 adapter
  - The short-hand method added by 574234b.
  - But only added in `NATIVE_DATABASE_TYPES` for PG adapter.

This PR is follow up to the above changes.

kamipo committed:
Add missing `serial` type in `NATIVE_DATABASE_TYPES` for PG adapter

The short-hand methods was added by #19030.
schema dumping support was added by #19185.

This commit is follow up to these PRs.

kaspth commented:
Don't think I'm the best reviewer for this 😁

R+ @rails-bot

senny commented:
@kamipo what is not working as long as this entry is missing? Can specify that in a test-case?

kaspth commented:
Be sure to clarify that in your commit message as well :grin:

kamipo committed:
Add missing geometric types in `NATIVE_DATABASE_TYPES` for PG adapter

The short-hand methods was added by #20459.

This commit is follow up to the PR.

kamipo committed:
Add missing `bigint` type in `NATIVE_DATABASE_TYPES` for SQLite3 adapter

The short-hand method added by 574234be067f0f000a945560cabcabb68c8a1ec9.
But only added in `NATIVE_DATABASE_TYPES` for PG adapter.

This commit is follow up to the commit.

kamipo commented:
This PR is nitpick.

`NATIVE_DATABASE_TYPES` is used in `type_to_sql` for look up to native database type.

```ruby
      def type_to_sql(type, limit = nil, precision = nil, scale = nil) #:nodoc:
        if native = native_database_types[type.to_sym]
          column_type_sql = (native.is_a?(Hash) ? native[:name] : native).dup
```

https://github.com/kamipo/rails/blob/add_missing_serial_in_native_database_types/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L999

In most case, not have any problem though forgot the additions to the `NATIVE_DATABASE_TYPES`.

It is because if missing in the `NATIVE_DATABASE_TYPES`, `type_to_sql` will `to_s` the `type` simply.

https://github.com/kamipo/rails/blob/add_missing_serial_in_native_database_types/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L1027

senny commented:
@kamipo but shouldn't the default behavior `to_s` be equivalent to adding all those entries? Does this change any actual behavior?

kamipo commented:
Ideally, the supported types should be added to the `NATIVE_DATABASE_TYPES`. It is because a column type is checked by `valid_type?` in `SchemaDumper`.

```ruby
          # then dump all non-primary key columns
          column_specs = columns.map do |column|
            raise StandardError, "Unknown type '#{column.sql_type}' for column '#{column.name}'" unless @connection.valid_type?(column.type)
            next if column.name == pk
            @connection.column_spec(column)
          end.compact
```

https://github.com/kamipo/rails/blob/add_missing_serial_in_native_database_types/activerecord/lib/active_record/schema_dumper.rb#L148

`valid_type?`:

```ruby
      def valid_type?(type)
        !native_database_types[type].nil?
      end
```

https://github.com/kamipo/rails/blob/add_missing_serial_in_native_database_types/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L373-L375

senny commented on activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:
```diff
@@ -62,7 +62,8 @@ class SQLite3Adapter < AbstractAdapter
         binary:       { name: "blob" },
-        boolean:      { name: "boolean" }
+        boolean:      { name: "boolean" },
+        bigint:       { name: "bigint" },
```
The sqlite3 adapter has no `valid_type?` method. Why would we need to add this entry?

kamipo replied:
Currently, `:bigint` and `:bigserial` entries does not affects the behavior because the type of these native column is an `:integer`.

My hope is that avoiding inconsistency.

Should we remove `:bigint` and `:bigserial` entries?

senny replied:
My guess is that we can probably remove most of the entries in the `Sqlite3Adapter` but let's explore that in another PR. As long as they don't make a transformation or have some non standard options, they won't affect behavior.

In terms of consistency. In my opinion we should target removing instead of adding new entries. Especially in sqlite3 where the column types don't really matter.

kamipo replied:
I see. I agree about removing the entries that won't affect behavior.

senny commented on activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:
```diff
@@ -102,6 +103,12 @@ class PostgreSQLAdapter < AbstractAdapter
+        box:         { name: "box" },
+        path:        { name: "path" },
+        polygon:     { name: "polygon" },
+        circle:      { name: "circle" },
```
As long as we have `valid_type?` implemented the way it is now, we should probably add those for consistency. Although my guess is that most people using these kind of data types are probably relying on `db/structure.sql` already.

kamipo replied:
Okay, I'll implement support for these geometric types dumping.

kamipo replied:
Implemented in #22304.

kamipo commented:
Closed in favor of #22302 and #22304. Thanks!
